presto-orc/src/main/java/com/facebook/presto/orc/OrcReader.java

```diff
@@ -186,6 +215,68 @@ public OrcReader(
         }
 
         this.cacheable = requireNonNull(cacheable, "hiveFileContext is null");
+
+        Optional<DwrfStripeCache> dwrfStripeCache = Optional.empty();
+        if (orcFileTail.getDwrfStripeCacheData().isPresent() && footer.getDwrfStripeCacheOffsets().isPresent()) {
+            DwrfStripeCache cache = buildDwrfStripeCache(footer.getStripes(), orcFileTail.getDwrfStripeCacheData().get(), footer.getDwrfStripeCacheOffsets().get());
+            dwrfStripeCache = Optional.of(cache);
+        }
+
+        requireNonNull(stripeMetadataSourceFactory, "stripeMetadataSourceFactory is null");
+        this.stripeMetadataSource = requireNonNull(stripeMetadataSourceFactory.create(dwrfStripeCache), "stripeMetadataSource is null");
+    }
+
+    private DwrfStripeCache buildDwrfStripeCache(List<StripeInformation> stripes, DwrfStripeCacheData dwrfStripeCacheData, List<Integer> cacheOffsets)
+    {
+        // AbstractOrcRecordReader sorts stripes by position, so probably there
+        // is a use-case when stripes can come unordered
+        List<StripeInformation> sortedStripes = new ArrayList<>(stripes);
+        sortedStripes.sort(comparingLong(StripeInformation::getOffset));
+        DwrfStripeCacheMode mode = dwrfStripeCacheData.getDwrfStripeCacheMode();
+        Slice cacheSlice = dwrfStripeCacheData.getDwrfStripeCacheSlice();
+
+        ImmutableMap.Builder<Long, Slice> indexSlices = ImmutableMap.builder();
+        if (mode.hasIndex()) {
+            for (int i = 0; i < sortedStripes.size(); i++) {
+                int offsetIdx = mode.getIndexOffsetPosition(i);
+                if (!addDwrfStripeCacheSlice(offsetIdx, cacheOffsets, cacheSlice, indexSlices, sortedStripes.get(i))) {
```

Comment:
Can we extract the valid check from `addDwrfStripeCacheSlice` to a separate method `isValidOffset`? This way I feel is a little bit cleaner that we don't need to return `boolean` for an adding action, and the logic in this for loop would become:

```
for (...) {
  if (!isValidOffset(offset)) {
    break;
  }
  addDwrfStripeCacheSlice(...);
}
```

Reply:
done